src/main/kotlin/com/exactpro/th2/codec/EventProcessor.kt

```diff
@@ -30,91 +30,141 @@ import com.exactpro.th2.common.utils.message.sessionAlias
-        parentEventIDs.forEach {
-            it.addReferenceTo(warnEvent, message, Event.Status.PASSED, messagesIds, additionalBody = body)
-        }
-    }
+        cause: Throwable? = null
+    ): EventID = (eventId ?: codecEventID).onErrorEvent(message, messagesIds, cause)
 
     fun onEachErrorEvent(
-        parentEventIDs: Sequence<EventID>,
+        pairIds: Map<MessageID, EventID?>,
         message: String,
-        messagesIds: List<MessageID> = emptyList(),
         cause: Throwable? = null,
-        additionalBody: List<String> = emptyList(),
-    ): EventID {
-        val errorEventId = codecEventID.onErrorEvent(message, messagesIds, cause, additionalBody)
-        parentEventIDs.forEach {
-            it.addReferenceTo(errorEventId, message, Event.Status.FAILED, messagesIds, cause, additionalBody)
+        body: List<String> = emptyList(),
+    ): Map<String, EventID> {
+        val bookToEventId = mutableMapOf<String, EventID>()
+        chooseRootEventsForPublication(pairIds).forEach { (codecEventId, externalEventIds) ->
+            val errorEventId = codecEventId.onErrorEvent(message, pairIds.keys, cause, body)
+            bookToEventId.put(errorEventId.bookName, errorEventId)?.also {
+                error(
+                    "Internal error: several root events have been chosen for pair ids: $pairIds"
+                )
+            }
+            externalEventIds.forEach {
+                it.addReferenceTo(errorEventId, message, Event.Status.FAILED, pairIds.keys, cause, body)
+            }
         }
-        return errorEventId
+        return bookToEventId
     }
 
-    fun onEachWarning(
-        parentEventIDs: Sequence<EventID>,
-        context: ReportingContext,
-        action: String,
-        additionalBody: () -> List<String> = ::emptyList,
-        messagesIds: () -> List<MessageID> = ::emptyList
-    ) = context.warnings.let { warnings ->
-        if (warnings.isNotEmpty()) {
-            val messages = messagesIds()
-            val body = additionalBody()
-            warnings.forEach { warning ->
-                onEachEvent(parentEventIDs, "[WARNING] During $action: $warning", messages, body)
+    internal fun chooseRootEventsForPublication(
+        pairIds: Map<MessageID, EventID?>,
+    ): Map<EventID, Set<EventID>> = mutableMapOf<EventID, MutableSet<EventID>>().apply {
+        pairIds.forEach { (messageId, eventId) ->
+            if (eventId == null) {
+                if (messageId.isValid) {
+                    compute(getRootEvent(messageId.bookName)) { _, value ->
+                        value ?: mutableSetOf()
+                    }
+                }
+            } else {
+                compute(getRootEvent(eventId.bookName)) { _, value ->
+                    (value ?: mutableSetOf()).apply { add(eventId) }
+                }
             }
         }
+
+        if (isEmpty()) {
+            put(codecEventID, Collections.emptySet())
```

Comment:
Kotlin also has `emptySet` method. It can be used instead of the method from `Collections` class

Reply:
Kotlin's emptySet method return Set interface instead of MutableSet.

Reply:
Okay, but this is very strange for kotlin to use a `Mutable*` interface for values and put there something that is not mutable... But okay, I got the point